### ISSUE TITLE
Add ability to sort SolutionArray

### ISSUE THREAD

**Issue by bryanwweber** [Cantera/cantera#625]
Right now, it is not possible to sort `SolutionArray` instances without copying everything into a new `SolutionArray` with a loop. Example, courtesy of @arghdos:

```python
import cantera as ct
gas = ct.Solution('gri30.cti')
arry = ct.SolutionArray(gas, extra=['tau'])
arry.append(gas.state, tau=1)
gas.TP = 3000, 101325
arry.append(gas.state, tau=0.5)
tau_idx = np.argsort(arry.tau)
out = ct.SolutionArray(gas, extra=['tau'])
for idx in tau_idx:
	out.append(arry._states[idx], tau=arry.tau[idx])
```

It would be useful to add a `.sort()` method that takes a column name or index as the argument and sorts based on that. To implement this will require converting the "internal" `_states` object to be consistently either a list or an array. The code below indicates (to me, anyways) that if the user provides the `states` argument, it is passed through as is (list or array), and if the shape of the `states` is going to be one-dimensional, the `_states` is a list.

https://github.com/Cantera/cantera/blob/f078f53f39118416e4d3349fce99b8e105d5b0b4/interfaces/cython/cantera/composite.py#L449-L459

**Comment by ischoegl** [comment]
@bryanwweber ... depending on what context `.sort()` is needed (during a processing step or before plotting), converting a `SolutionArray` to a `pandas.DataFrame` may provide a simple work-around.

I.e. if no additional `SolutionArray` manipulations are needed, this would work: 
```python
import pandas as pd

# convert solution array states (with extra being 'tau')
data, labels = states.collect_data(cols=('extra','T','density','Y'))
df = pd.DataFrame(data=data, columns=labels)
df.sort_values(by='tau', inplace=True)
```
or
```python
# using 'extra' as index
df = pd.DataFrame(data=data[:,1:], columns=labels[1:],index=data[:,0])
df.sort_index(inplace=True)
```
As there is currently no C++ implementation of `SolutionArray` objects, the python data analysis package [`pandas`](https://pandas.pydata.org/) does not restrict usage, and brings a considerable amount of functionality along (slicing, access of columns as attributes, file input/output, etc.).

**Comment by ischoegl** [comment]
>
> To implement this will require converting the "internal" `_states` object to be consistently either a list or an array. The code below indicates (to me, anyways) that if the user provides the `states` argument, it is passed through as is (list or array), and if the shape of the `states` is going to be one-dimensional, the `_states` is a list.

In my PR I did not implement sorting for higher-dimensional `SolutionArray` objects as a good definition is not obvious to me (it *is* possible, but I cannot think of a good reason).

It is true that 1D and higher dimensional `SolutionArray` object are implemented differently. My suspicion is that the rationale for this implementation is that in 1D memory can be accessed directly, whereas in higher dim's copying cannot be avoided. I also added a simple type check for the optional `states` argument in PR #688. This is used in `__call__` (a copying constructor?), but I haven't seen anything else using this yet.
